common-custom-user-data-gradle-plugin/src/main/java/com/gradle/CustomBuildScanEnhancements.java

```diff
@@ -27,67 +29,59 @@
  */
 final class CustomBuildScanEnhancements {
 
-    static void configureBuildScan(BuildScanExtension buildScan, Gradle gradle) {
-        captureOs(buildScan);
-        captureIde(buildScan, gradle);
-        captureCiOrLocal(buildScan);
-        captureCiMetadata(buildScan, gradle);
+    static void configureBuildScan(BuildScanExtension buildScan, Gradle gradle, ProviderFactory providers) {
+        captureOs(buildScan, providers);
+        captureIde(buildScan, gradle, providers);
+        captureCiOrLocal(buildScan, providers);
+        captureCiMetadata(buildScan, gradle, providers);
         captureGitMetadata(buildScan);
         captureTestParallelization(buildScan, gradle);
     }
 
-    private static Optional<String> projectProperty(Gradle gradle, String name) {
-        String value = (String) gradle.getRootProject().findProperty(name);
-        return Optional.ofNullable(value);
+    private static void captureOs(BuildScanExtension buildScan, ProviderFactory providers) {
+        sysProperty("os.name", providers).ifPresent(buildScan::tag);
     }
 
-    private static void captureOs(BuildScanExtension buildScan) {
-        sysProperty("os.name").ifPresent(buildScan::tag);
-    }
-
-    private static void captureIde(BuildScanExtension buildScan, Gradle gradle) {
+    private static void captureIde(BuildScanExtension buildScan, Gradle gradle, ProviderFactory providers) {
```

Comment:
I believe you can refactor these if/else into ifPresent

Reply:
It is possible, but getting the same semantics is tricky. I'm not sure the `ifPresent` version is better/clearer than the `if-else if` version. Here's the best that I could come up with:

```java
final AtomicReference<String> ideTag = new AtomicReference<>("");
projectProperty("android.injected.invoked.from.ide", providers, gradle).ifPresent(__ -> ideTag.set("Android Studio"));
projectProperty("android.injected.studio.version", providers, gradle).ifPresent(v -> buildScan.value("Android Studio version", v));

sysProperty("idea.version", providers).ifPresent(__ -> ideTag.set("IntelliJ IDEA"));
sysProperty("eclipse.buildId", providers).ifPresent(__ -> ideTag.set("Eclipse"));

if(sysPropertyKeyStartingWith("idea.version")) {
    ideTag.set("IntelliJ IDEA");
}
if(!isCi(providers)) {
    ideTag.set("Cmd Line");
}

if (!ideTag.get().isEmpty()) {
    buildScan.tag(ideTag.get());
}
```

If we think it's better, then I'll gladly change the code to the above. Other suggestions are also welcome.

I thought about using `Optional.ifPresentOrElse`, but that is even more complex, and doesn't work for every case since some of the else if clauses contain regular boolean expressions (like `!isCi(providers`).

Reply:
Let's leave the current if/else structure.